src/desc.rs

```diff
@@ -0,0 +1,146 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, HashSet};
+use std::hash::Hasher;
+
+use fnv::FnvHasher;
+
+use proto::LabelPair;
+use errors::{Result, Error};
+use metrics::SEPARATOR_BYTE;
+
+// Desc is the descriptor used by every Prometheus Metric. It is essentially
+// the immutable meta-data of a Metric. The normal Metric implementations
+// included in this package manage their Desc under the hood. Users only have to
+// deal with Desc if they use advanced features like the ExpvarCollector or
+// custom Collectors and Metrics.
+//
+// Descriptors registered with the same registry have to fulfill certain
+// consistency and uniqueness criteria if they share the same fully-qualified
+// name: They must have the same help string and the same label names (aka label
+// dimensions) in each, constLabels and variableLabels, but they must differ in
+// the values of the constLabels.
+//
+// Descriptors that share the same fully-qualified names and the same label
+// values of their constLabels are considered equal.
+//
+// Use NewDesc to create new Desc instances.
+#[derive(Clone, Debug)]
+pub struct Desc {
+    // fq_name has been built from Namespace, Subsystem, and Name.
+    pub fq_name: String,
+    // help provides some helpful information about this metric.
+    pub help: String,
+    // const_label_pairs contains precalculated DTO label pairs based on
+    // the constant labels.
+    pub const_label_pairs: Vec<LabelPair>,
+    // variable_labels contains names of labels for which the metric
+    // maintains variable values.
+    pub variable_labels: Vec<String>,
+    // id is a hash of the values of the ConstLabels and fqName. This
+    // must be unique among all registered descriptors and can therefore be
+    // used as an identifier of the descriptor.
+    pub id: u64,
+    // dim_hash is a hash of the label names (preset and variable) and the
+    // Help string. Each Desc with the same fqName must have the same
+    // dimHash.
+    pub dim_hash: u64,
+}
+
+impl Desc {
+    pub fn new(fq_name: String,
+               help: String,
+               variable_labels: Vec<String>,
+               const_labels: HashMap<String, String>)
+               -> Result<Desc> {
+        let mut desc = Desc {
+            fq_name: fq_name.clone(),
+            help: help,
+            const_label_pairs: Vec::with_capacity(const_labels.len()),
+            variable_labels: vec![],
+            id: 0,
+            dim_hash: 0,
+        };
+
+        if desc.help.is_empty() {
+            return Err(Error::Msg("empty help string".into()));
+        }
+
+        // TODO: check valid metric fq_name.
+
+        let mut label_values = Vec::with_capacity(const_labels.len() + 1);
+        label_values.push(fq_name);
+        let mut label_names = Vec::with_capacity(const_labels.len() + variable_labels.len());
```

Comment:
Just port client go.

Let's keep it here, I will refactor all when we introduce label support.